File: java/users/Lookup.java
/**
 * User Lookup - X API v2
 * 
 * Endpoint: GET https://api.x.com/2/users/by
 * Docs: https://developer.x.com/en/docs/twitter-api/users/lookup/api-reference/get-users-by
 * 
 * Authentication: Bearer Token (App-only)
 * Required env vars: BEARER_TOKEN
 * 
 * Dependencies: org.apache.httpcomponents:httpclient, com.google.code.gson:gson
 */

import java.io.IOException;
import java.net.URISyntaxException;
import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.client.HttpClient;
import org.apache.http.client.config.CookieSpecs;
import org.apache.http.client.config.RequestConfig;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.utils.URIBuilder;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.util.EntityUtils;

public class Lookup {

    public static void main(String[] args) throws IOException, URISyntaxException {
        String bearerToken = System.getenv("BEARER_TOKEN");
        
        if (bearerToken == null) {
            System.err.println("BEARER_TOKEN environment variable not set");
            System.exit(1);
        }

        String response = lookupUsers("XDevelopers,X", bearerToken);
        System.out.println(response);
    }

    private static String lookupUsers(String usernames, String bearerToken) throws IOException, URISyntaxException {
        String userLookupUrl = "https://api.x.com/2/users/by";

        HttpClient httpClient = HttpClients.custom()
                .setDefaultRequestConfig(RequestConfig.custom()
                        .setCookieSpec(CookieSpecs.STANDARD).build())
                .build();

        URIBuilder uriBuilder = new URIBuilder(userLookupUrl);
        uriBuilder.addParameter("usernames", usernames);
        uriBuilder.addParameter("user.fields", "created_at,description,public_metrics");

        HttpGet httpGet = new HttpGet(uriBuilder.build());
        httpGet.setHeader("Authorization", "Bearer " + bearerToken);
        httpGet.setHeader("User-Agent", "v2UserLookupJava");

        HttpResponse response = httpClient.execute(httpGet);
        HttpEntity entity = response.getEntity();
        
        if (entity != null) {
            return EntityUtils.toString(entity, "UTF-8");
        } else {
            return "No response";
        }
    }
}

File: python/users/lookup.py
"""
User Lookup - X API v2
======================
Endpoint: GET https://api.x.com/2/users/by
Docs: https://developer.x.com/en/docs/twitter-api/users/lookup/api-reference/get-users-by

Authentication: Bearer Token (App-only) or OAuth (User Context)
Required env vars: BEARER_TOKEN
"""

import requests
import os
import json

bearer_token = os.environ.get("BEARER_TOKEN")


def create_url():
    # Specify the usernames to lookup (up to 100 comma-separated)
    usernames = "usernames=XDevelopers,X"
    
    # User fields are adjustable. Options include:
    # created_at, description, entities, id, location, name,
    # pinned_tweet_id, profile_image_url, protected,
    # public_metrics, url, username, verified, and withheld
    user_fields = "user.fields=description,created_at"
    url = "https://api.x.com/2/users/by?{}&{}".format(usernames, user_fields)
    return url


def bearer_oauth(r):
    """
    Method required by bearer token authentication.
    """
    r.headers["Authorization"] = f"Bearer {bearer_token}"
    r.headers["User-Agent"] = "v2UserLookupPython"
    return r


def connect_to_endpoint(url):
    response = requests.request("GET", url, auth=bearer_oauth)
    print(response.status_code)
    if response.status_code != 200:
        raise Exception(
            "Request returned an error: {} {}".format(
                response.status_code, response.text
            )
        )
    return response.json()


def main():
    url = create_url()
    json_response = connect_to_endpoint(url)
    print(json.dumps(json_response, indent=4, sort_keys=True))


if __name__ == "__main__":
    main()
